File: Contextual/src/contextual/ContextualRerank.java
package contextual;

import util.LoadMap;
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Properties;

import com.mindprod.ledatastream.LEDataInputStream;
import java.util.Arrays;

public class ContextualRerank {

    private static String DESCRIPTORS_MAP;
    private static String COLLECTION_MAP;

    private static final String EXTENSION = ".distbin";

    private static Properties descriptorProperties;
    private static Properties collectionProperties;

    private static final int COLLECTION_SIZE = 20000;

    public static void main(String[] args) throws FileNotFoundException, IOException {
        ContextualRerank calc = new ContextualRerank("/home/luciano/Desktop/ic/Rodrigo/descriptors/", "src/resources/clef.map");
        calc.run();
        long time = System.currentTimeMillis();
        calc.calculateContext(2, 2);
        System.out.println("Tempo total:\t\t" + String.format("%.2f", (System.currentTimeMillis() - time) / 1000.0 / 3600.0));
        System.out.println("ok");
    }

    public ContextualRerank(String descriptorsDir, String collectionDir) {
        this.DESCRIPTORS_MAP = descriptorsDir + "descriptors.map";
        this.COLLECTION_MAP = collectionDir;
    }

    /**
     * Runs class
     */
    public void run() {
        setUp();
    }

    /**
     * Build an image's KNN given its path and its amount of neighbors. Obs:
     * Both images that we are using to compute distances has to be filtered out
     * from the set, considering that we can't take into account the distances
     * to themselves.
     *
     * @param imgRankPath the path from an image rank
     * @param K its amount of neighbors
     * @return an array with the ID's from an image's KNN in decreasing order
     */
    private int[] buildKNN(String imgRankPath, int K, int filter) {
        int[] knn = new int[K];
        try {
            FileReader reader = new FileReader(imgRankPath);
            BufferedReader buffer = new BufferedReader(reader);
            buffer.readLine(); // discard the distance to itself
            for (int i = 0; i < knn.length; i += 1) {
                int n = Integer.parseInt(buffer.readLine());
                if (n == filter) {
                    n = Integer.parseInt(buffer.readLine());
                }
                knn[i] = n;
            }
        } catch (FileNotFoundException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        } catch (IOException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
        return knn;
    }

    /**
     * Generate a reverse map for a given properties map.
     */
    public static void reverseMap(Properties p, String collectionDir) {
        File file = new File(collectionDir + ".rmap");
        try {
            BufferedWriter bw = new BufferedWriter(new FileWriter(file));
            for (int j = 0; j < p.size(); j++) {
                bw.write(p.getProperty(Integer.toString(j)) + "=" + j + "\n");
            }
            bw.close();
        } catch (FileNotFoundException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        } catch (IOException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
    }

    /**
     * Calculate contextual distances for the given K values. Obs: The starter
     * value associated to Ks has to be bigger than 2 and the ending value Ke
     * has to be lesser than (N-2) being N the size of the collection given the
     * fact that when new distances are computed and a set of K nearest
     * neighbors is built, two values are disregarded (two images that we are
     * using to compute contextual distances)
     * @param Ks
     * @param Ke
     */
    public void calculateContext(int Ks, int Ke) {
        int K = Ks;
        while (K <= Ke) {
            contextualRerank(K++);
        }
    }

    /**
     * Calculate new distances utilizing contextual re-rank.
     */
    private void contextualRerank(int K) {
        String descriptorPath = this.DESCRIPTORS_MAP.substring(0, this.DESCRIPTORS_MAP.length() - 16); // descriptor path		
        for (int d = 1; d < this.descriptorProperties.size() - 10; d += 1) { // for each descriptor D
            String descriptor = descriptorProperties.getProperty(Integer.toString(d));// gets the descriptor name
            System.out.println(descriptor);
            System.out.printf("Calculando contextos usando K = %d %n", K);
            for (int l = 20117; l < 20118; l += 1) { // for each imgL(topics) do:
                double[] contextualDistances = new double[COLLECTION_SIZE];// collection size			
                String imgL = descriptorPath + "/" + descriptor + "/" + collectionProperties.getProperty(Integer.toString(l)) + EXTENSION; // get imgL's path				
                System.out.println(imgL);
                for (int i = 0; i < COLLECTION_SIZE; i++) { // for each imgI(collection) do: compute distance between L & I using contextual information
                    if (i != l) {
                        String imgI = descriptorPath + "/" + descriptor + "/" + collectionProperties.getProperty(Integer.toString(i)) + EXTENSION; // get imgI's path
                        System.out.println(imgI);
                        int[] knn = buildKNN(imgI.substring(0, imgI.length() - 12) + ".rank", K, l); // build imgI's KNN
                        int ck = 0;
                        double dj = 0;
                        for (int j : knn) {// for each imgJ((KNN)I) do : weighted sum of distance from imgI neighbors, to imgL												
                            dj = dj + dist(j, l, d) * (K - ck);
                            ck += 1;
                        }
                        double di = dist(i, l, d) / K;
                        dj = dj / (K * (K + 1) / 2);
                        di = Math.pow(di, 2);
                        dj = Math.pow(dj, 2);
                        contextualDistances[i] = Math.sqrt(di + dj); // recalculate distance from imgI to imgL
                        System.out.println(contextualDistances[i]);
                    }
                }
                String path = imgL.substring(0, imgL.length() - 12) + ".cs_" + Integer.toString(K);
//				write(contextualDistances,path);
                if (l % 10 == 0) // show the progress in a scale of 10 
                {
                    System.out.println("[" + "Distancias recalculadas: " + (l - COLLECTION_SIZE) + "/" + 180 + ", descritor " + d + "/" + descriptorProperties.size() + "]");
                }
            }
            System.out.println("[" + "Descritor pronto: " + descriptorProperties.getProperty(Integer.toString(d)) + "]");
        }
    }

    /**
     * Returns the distance between two images for a given descriptor: Opens the
     * img1 rank in the corresponding position to distance for img2 and reads
     * it. TODO refactor verify if changing between reading file libraries is
     * really necessary.
     */
    private static double dist(int img1, int img2, int descriptor) {
        String descriptorPath = DESCRIPTORS_MAP.substring(0, DESCRIPTORS_MAP.length() - 16); // descriptor path		
        String descriptorName = descriptorProperties.getProperty(Integer.toString(descriptor));
        double val = 0;
        try {
            String imgJ = descriptorPath + "/" + descriptorName + "/" + collectionProperties.getProperty(Integer.toString(img1));
            FileInputStream fis = new FileInputStream(imgJ + EXTENSION); // load the rank from imgJ
            fis.skip(8 * img2);
//						DataInputStream lis = new DataInputStream(fis);
            LEDataInputStream lis = new LEDataInputStream(fis); // usar caso little ending
            val = lis.readDouble();
            lis.close();
            fis.close();
        } catch (FileNotFoundException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        } catch (IOException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
        return val;
    }

    /**
     * Write an double array to a respective path. TODO write the array
     */
    private static void write(double[] distances, String path) {

        File file;
        file = new File(path);
        try {
            BufferedWriter bw = new BufferedWriter(new FileWriter(file));
            for (int i = 0; i < distances.length; i += 1) {
                bw.write(distances[i] + "\n");
            }
            bw.close();
        } catch (FileNotFoundException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        } catch (IOException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
    }

    /**
     * Sets up maps TODO refactor, LoadMap class is not needed.
     */
    private void setUp() {
        LoadMap load = new LoadMap(this.DESCRIPTORS_MAP); // loads a map containing the descriptors
        load.load();
        descriptorProperties = load.export();
        load = new LoadMap(this.COLLECTION_MAP); // loads a map containing the collection
        load.load();
        collectionProperties = load.export();
    }
}

File: v2/src/contextual_rerank.c
#include <stdio.h>
#include <stdlib.h>
#include "rank.c"
#include "math.h"

/**
*   Build an image's kNN given its amount of neightbors.

*   Params: 
*   K: amount of neighbors
*   filter: Since that we are computing the knn of an image regarding
*   an other image, we need to filter both of them from the kNN set
*   that is going to be built. Therefore 
**/
int *buildKNN(int K, int imgI, int imgJ, Rank r)
{
    int *kNN = malloc(K * sizeof(int));
    int pos = 0;
    int aux = 0;
    if (r.entries[aux].id == imgI)
    {
        aux++;
    }
    while (pos < K)
    {
        int tmp = r.entries[aux].id;
        if (tmp == imgJ)
        {
            aux++;
        }
        kNN[pos++] = r.entries[aux++].id;
    }
    return kNN;
}

Rank *rerank(Distbin distances[20180])
{
    Rank *ranks = malloc(180 * sizeof(Rank));
    for (int i = 0; i < 180; i++)
    {
        ranks[i] = NewRank(&distances[i]);
    }
    return ranks;
}

int *contextual_rerank(Distbin a[20180], Rank *r, int K)
{
    Distbin *a_line = malloc(180 * sizeof(Distbin));
    for (int i = 0; i < 180; i++)
    {
        for (int j = 0; j < 20180; j++)
        {
            a_line[i].distances[j] = a[i].distances[j];
        }
    }

    for (int imgI = 20000; imgI < 20180; imgI++)
    { // for each imgI
        for (int imgL = 0; imgL < 20000; imgL++)
        { // for each imgL (topics rows)
            if (imgI != imgL)
            { // discard itself
                Rank rank = r[imgI - 20000];
                int *kNN = buildKNN(K, imgI, imgL, rank);
                int ck = 0;
                double dj = 0;
                for (int j = 0; j < K; j++)
                { // for each imgJ((KNN)L) do : weighted sum of distance from imgI neighbors, to imgL
                    int imgJ = kNN[j];
                    dj = dj + a[imgJ].distances[imgL] * (K - ck);
                    ck++;
                }
                double di = a[imgI].distances[imgL] / K;
                dj = dj / (K * (K + 1) / 2);
                di = di * di;
                dj = dj * dj;
                a_line[imgI - 20000].distances[imgL] = sqrt(di + dj); // recalculate distance from imgI to imgL
            }
        }
    }

    for (int i = 20000; i < 20180; i++) //Overwrite previous values for the original distances
    {
        for (int j = 0; j < 20180; j++)
        {
            a[i].distances[j] = a_line[i - 20000].distances[j];
        }
    }
    return a_line;
}

void ContextualRerank(Distbin original_distances[20180], int Ks, int Ke)
{
    int K = Ks;
    while (K < Ke)
    {
        Rank *ranked_lists = rerank(original_distances);
        original_distances = contextual_rerank(original_distances, ranked_lists, K);
        K++;
    }
}

int main(int argc, char const *argv[])
{
    printf("Sizeof argv: %d\n", argc);
    if (sizeof(argv) < 2)
    {
        printf("Usage: ./contextual [descriptor_id] [Ks] [Ke]\n");
    }
    else
    {
        int arg0 = atoi(argv[1]);
        int arg1 = atoi(argv[2]);
        int arg2 = atoi(argv[3]);
    }

    FILE *pointer;
    pointer = fopen("../../Contextual/resources/images/16672.jpg.ppm.distbin", "r");
    if (!pointer)
    {
        printf("Unable to access \n");
    }
    else
    {
        Distbin *distbin = NewDistbin(pointer, distbin);
        printf("˜here we go again˜\n");
        Rank r = NewRank(distbin);
        for (int i = 0; i < 8; i++)
        {
            printf("Distance:%lf\n", r.entries[i].distance);
            printf("Id: %d\n", r.entries[i].id);
        }
        int *knn = buildKNN(7, 16672, 9739, r);
        for (int i = 0; i < 7; i++)
        {
            printf("Id knn[%d](img 16672): %d\n", i, knn[i]);
        }
    }
    return 0;
}